### PR TITLE
Feature/restructure outputs

### PR DESCRIPTION
**Pull request by KateSakharova**
**Updates:**
I added FASTQC for decontaminated reads. Renamed files in decontamination step very urgly...

MultiQC report example. **Have a look!**
[multiqc_report.html.zip](https://github.com/EBI-Metagenomics/miassembler/files/15376013/multiqc_report.html.zip)

**Current structure:**
```
├── assembly
│   ├── decontamination
│   │   └── SRR6180434.txt
│   ├── spades
│   │   └── 3.15.5
│   │       ├── params.txt
│   │       ├── SRR6180434.assembly_graph.fastg.gz
│   │       ├── SRR6180434.assembly_graph_with_scaffolds.gfa.gz
│   │       ├── SRR6180434.contigs.fa.gz
│   │       └── SRR6180434.scaffolds.fa.gz
|   |       └── coverage
│   │              └── SRR6180434.txt.gz
|   |       └── qc
│   |          ├── multiqc
│   |          │   ├── multiqc_data
│   |          │   ├── multiqc_plots
│   |          │   └── multiqc_report.html
│   |          └── quast
│   |              ├── SRR6180434
│   |              └── versions.yml
├── pipeline_info
│   ├── execution_report_2024-05-17_16-17-51.html
│   ├── execution_timeline_2024-05-17_16-17-51.html
│   ├── execution_trace_2024-05-17_16-17-51.txt
│   ├── params_2024-05-17_16-18-58.json
│   ├── pipeline_dag_2024-05-17_16-17-51.html
│   └── software_versions.yml
└── qc
    ├── fastp
    │   ├── SRR6180434.fastp.html
    │   ├── SRR6180434.fastp.json
    │   └── SRR6180434.fastp.log
    └── fastqc
        ├── decontaminated_SRR6180434_1_fastqc.html
        ├── decontaminated_SRR6180434_1_fastqc.zip
        ├── decontaminated_SRR6180434_2_fastqc.html
        ├── decontaminated_SRR6180434_2_fastqc.zip
        ├── SRR6180434_1_fastqc.html
        ├── SRR6180434_1_fastqc.zip
        ├── SRR6180434_2_fastqc.html
        ├── SRR6180434_2_fastqc.zip
        └── versions.yml_fastqc.zip
        └── versions.yml

```
**Problems:**
I can output assembly into assembler/version, but I can't easily add coverage/qc/decontamination into that folder. For assembly process we do not refer to params.assembler because we have this lovely bit:
```
    READS_QC.out.qc_reads.branch { meta, reads ->
        xspades: ["metaspades", "spades"].contains(params.assembler)
                && meta.single_end == false
                || isMetatranscriptomic
        megahit: params.assembler == "megahit" || meta.single_end == true
    }.set { qc_reads }
```
and if I use `$params.assembler / coverage` (where requested assembler = spades) but during pipeline assembler was changed to megahit - then coverage would be in `spades` but assembly results would be in `megahit` (because that process was launched).

TODO:
- spades vs metaspades? how to change folder name?

### CODE REVIEW

File: conf/modules.config [hunk 1]
```diff
@@ -59,13 +85,47 @@ process {
...
+                mode: params.publish_dir_mode,
+                failOnError: true,
+                pattern: "*scaffolds.fa.gz"
+            ],
+            [
+                path: "${params.outdir}/assembly/spades/${params.spades_version}",
+                mode: params.publish_dir_mode,
+                failOnError: true,
+                pattern: "*graph_with_scaffolds.gfa.gz"
+            ]
```
Comment (mberacochea) [inline conf/modules.config:127]: I think this should work (less repetition)
```suggestion
            [
                path: "${params.outdir}/assembly/spades/${params.spades_version}",
                mode: params.publish_dir_mode,
                failOnError: true,
                saveAs: {
                    if ( filename.equals('versions.yml') ) {
                        return null;
                    }
                    def file_patterns_to_publish = [
                        "params.txt",
                        "graph.fastg",
                        "contigs.fa.gz",
                        "scaffolds.fa.gz",
                        "graph_with_scaffolds.gfa.gz"
                    ]
                    if ( file_patterns_to_publish.any { it.contains(filename) } ) {
                        return filename;
                    }
                    return null;
                }
            ]
```

File: modules/ebi-metagenomics/samtools/bam2fq/main.nf [hunk 2]
```diff
@@ -21,7 +21,7 @@ process SAMTOOLS_BAM2FQ {
 
     script:
     def args = task.ext.args ?: ''
-    def prefix = task.ext.prefix ?: "${meta.id}"
+    def prefix = task.ext.prefix ? "decontaminated_${meta.id}": "${meta.id}"
```
Comment (mberacochea) [inline modules/ebi-metagenomics/samtools/bam2fq/main.nf:24]: To keep things consistent, I would put this one in the `task.ext` in the modules just like BWA_MEM2 (https://github.com/EBI-Metagenomics/miassembler/pull/6/files#diff-bf809c928cb10b54d251dec8a140a2d5505150f97175d8d7b51dda9cc57971feR88)

**Review by mberacochea — COMMENTED** [review]
Thanks Kate. I left you some comments.
One thing that is missing is the study / run version on the folder name
```

SRPXXX / SRPXXXX
├── SRR1240
│     ├── SRR12403548
│     │    │   ├── assembly
│     │    │   ├── decontamination
│     │    │   │   └── SRR6180434.txt
│     │    │   ├── spades
│     │    │   │   └── 3.15.5
│     │    │   │       ├── params.txt
│     │    │   │       ├── SRR6180434.assembly_graph.fastg.gz
│     │    │   │       ├── SRR6180434.assembly_graph_with_scaffolds.gfa.gz
│     │    │   │       ├── SRR6180434.contigs.fa.gz
│     │    │   │       └── SRR6180434.scaffolds.fa.gz
|     |    |   |       └── coverage
│     │    │   │              └── SRR6180434.txt.gz
|     |    |   |       └── qc
│     │    │   |          ├── multiqc
│     │    │   |          │   ├── multiqc_data
│     │    │   |          │   ├── multiqc_plots
│     │    │   |          │   └── multiqc_report.html
│     │    │   |          └── quast
│     │    │   |              ├── SRR6180434
│     │    │   |              └── versions.yml
│     │    ├── pipeline_info
│     │    │   ├── execution_report_2024-05-17_16-17-51.html
│     │    │   ├── execution_timeline_2024-05-17_16-17-51.html
│     │    │   ├── execution_trace_2024-05-17_16-17-51.txt
│     │    │   ├── params_2024-05-17_16-18-58.json
│     │    │   ├── pipeline_dag_2024-05-17_16-17-51.html
│     │    │   └── software_versions.yml
│     │    └── qc
│     │        ├── fastp
│     │        │   ├── SRR6180434.fastp.html
│     │        │   ├── SRR6180434.fastp.json
│     |        │   └── SRR6180434.fastp.log
│     │        └── fastqc
│     |            ├── decontaminated_SRR6180434_1_fastqc.html
│     │            ├── decontaminated_SRR6180434_1_fastqc.zip
│     │            ├── decontaminated_SRR6180434_2_fastqc.html
│     │            ├── decontaminated_SRR6180434_2_fastqc.zip
│     │            ├── SRR6180434_1_fastqc.html
│     │            ├── SRR6180434_1_fastqc.zip
│     │            ├── SRR6180434_2_fastqc.html
│     │            ├── SRR6180434_2_fastqc.zip
│     │            └── versions.yml_fastqc.zip
│     │            └── versions.yml
```

The attached multiqc report looks very good, but it has an extra software section include only fastqc:
image

**Comment by mberacochea** [comment]
We also need to remove any mentions about this workflow being part of nf-core, such as the multiqc methods text:
```
Data was processed using ebi-metagenomics/miassembler v1.0dev of the nf-core collection of workflows ....
```

File: subworkflows/local/assembly_qc.nf [hunk 3]
```diff
@@ -54,12 +55,14 @@ workflow ASSEMBLY_QC {
...
-            .collectFile(storeDir: "${params.outdir}/decontamination/contaminated_contigs.txt", newLine: true)
+            .collectFile(name: "contaminated_contigs_host.txt", newLine: true)
     } else {
         contaminated_contigs = BLAST_BLASTN_HUMAN_PHIX.out.txt
     }
 
-    contaminated_contigs.map { _, hits_txt -> hits_txt }.collectFile(storeDir: "${params.outdir}/assembly_qc/contaminated_contigs.txt", newLine: true)
+    collected_contigs = contaminated_contigs.map { meta, hits_txt -> hits_txt }.collectFile(name: "decontaminated.txt", newLine: true)
+    output_path = contaminated_contigs.map { meta, _ -> "assembly/$meta.assembler/$meta.assembler_version/decontamination/"}
+    PUBLISH_FILE(collected_contigs, output_path)
```
[the remark this answers is not in the document]
Comment (KateSakharova) [reply subworkflows/local/assembly_qc.nf:65]: meta is unknown in your suggestion

File: modules/nf-core/megahit/main.nf [hunk 4]
```diff
@@ -7,6 +7,13 @@ process MEGAHIT {
         'https://depot.galaxyproject.org/singularity/mulled-v2-0f92c152b180c7cd39d9b0e6822f8c89ccb59c99:8ec213d21e5d03f9db54898a2baeaf8ec729b447-0' :
         'biocontainers/mulled-v2-0f92c152b180c7cd39d9b0e6822f8c89ccb59c99:8ec213d21e5d03f9db54898a2baeaf8ec729b447-0' }"
 
+    publishDir(
```
[the remark this answers is not in the document]
Comment (KateSakharova) [reply modules/nf-core/megahit/main.nf:10]: because of meta

**Review by mberacochea — APPROVED** [review]
:shipit: I think this is ready to be merge now

**Pull request merged**